File: recettes.pl
% Définition des recettes et leurs ingrédients
recipe(pizza, [farine, eau, sel, levure, tomate, fromage]).
recipe(salade, [laitue, tomate, concombre, vinaigre, huile]).
recipe(pates_carbonara, [pates, creme, lardons, fromage, sel, poivre]).
recipe(omelette, [oeufs, sel, poivre, fromage]).
recipe(sandwich, [pain, beurre, jambon]).

% Vérifie si tous les ingrédients requis sont disponibles
has_all_ingredients(Recipe, Available) :-
    recipe(Recipe, Required),
    subset(Required, Available).

% Trouve toutes les recettes possibles avec les ingrédients disponibles
find_possible_recipes(Available, Recipes) :-
    findall(Recipe, has_all_ingredients(Recipe, Available), Recipes).

% Exemples d'utilisation :
% ?- find_possible_recipes([farine, eau, sel, levure, tomate, fromage, lardons], Recipes).
% ?- find_possible_recipes([oeufs, sel, poivre, fromage, jambon], Recipes).

File: recettes.py
def find_recipes_imperative(available_ingredients):
    recipes = {
        'Pizza': ['farine', 'eau', 'sel', 'levure', 'tomate', 'fromage'],
        'Salade': ['laitue', 'tomate', 'concombre', 'vinaigre', 'huile'],
        'Pates Carbonara': ['pates', 'creme', 'lardons', 'fromage', 'sel', 'poivre'],
        'Omelette': ['oeufs', 'sel', 'poivre', 'fromage'],
        'Sandwich': ['pain', 'beurre', 'jambon']
    }
    possible_recipes = []
    for recipe_name, required_ingredients in recipes.items():
        can_make = True
        for ingredient in required_ingredients:
            if ingredient not in available_ingredients:
                can_make = False
                break
        if can_make:
            possible_recipes.append(recipe_name)
    return possible_recipes

# Tests du programme
test1 = ['farine', 'eau', 'sel', 'levure', 'tomate', 'fromage']
print("Test 1 - Ingrédients disponibles:", test1)
print("Recettes possibles:", find_recipes_imperative(test1))

test2 = ['farine', 'eau', 'sel', 'levure', 'tomate', 'fromage', 'oeufs', 'poivre']
print("\nTest 2 - Ingrédients disponibles:", test2)
print("Recettes possibles:", find_recipes_imperative(test2))

test3 = ['oeufs', 'sel', 'poivre', 'fromage']
print("\nTest 3 - Ingrédients disponibles:", test3)
print("Recettes possibles:", find_recipes_imperative(test3))